pytools/__init__.py

```diff
@@ -267,43 +271,72 @@ def delta(x, y):
         return 0
 
 
-def levi_civita(tup):
-    """Compute an entry of the Levi-Civita tensor for the indices *tuple*."""
+def levi_civita(tup: Tuple[int, ...]) -> int:
+    """Compute an entry of the Levi-Civita symbol for the indices *tuple*."""
     if len(tup) == 2:
         i, j = tup
-        return j-i
+        return j - i
     if len(tup) == 3:
         i, j, k = tup
-        return (j-i)*(k-i)*(k-j)/2
+        return (j-i) * (k-i) * (k-j) // 2
     else:
-        raise NotImplementedError
+        raise NotImplementedError(f"Levi-Civita symbol in {len(tup)} dimensions")
 
 
-def factorial(n):
-    from operator import mul
-    assert n == int(n)
-    return reduce(mul, (i for i in range(1, n+1)), 1)
+factorial = MovedFunctionDeprecationWrapper(math.factorial)
 
 
-def perm(n, k):
-    """Return P(n, k), the number of permutations of length k drawn from n
-    choices.
-    """
-    result = 1
-    assert k > 0
-    while k:
-        result *= n
-        n -= 1
-        k -= 1
+try:
+    # NOTE: only available in python > 3.8
+    perm = math.perm
+except AttributeError:
+    def perm(n: SupportsIndex, k: Optional[SupportsIndex] = None) -> int:
+        """
+        :returns: :math:`P(n, k)`, the number of permutations of length :math:`k`
+            drawn from :math:`n` choices.
+        """
+        if k is None:
+            return math.factorial(n)
 
-    return result
+        import operator
+        n, k = operator.index(n), operator.index(k)
+        if k > n:
+            return 0
 
+        if k < 0:
+            raise ValueError("k must be a non-negative integer")
 
-def comb(n, k):
-    """Return C(n, k), the number of combinations (subsets)
-    of length k drawn from n choices.
-    """
-    return perm(n, k)//factorial(k)
+        if n < 0:
+            raise ValueError("n must be a non-negative integer")
+
+        from numbers import Integral
+        if not isinstance(k, Integral):
+            raise TypeError(f"'{type(k).__name__}' object cannot be interpreted "
+                            "as an integer")
+
+        if not isinstance(n, Integral):
+            raise TypeError(f"'{type(n).__name__}' object cannot be interpreted "
+                            "as an integer")
+
+        result = 1
+        while k:
+            result *= n
+            n -= 1
+            k -= 1
+
+        return result
```

Comment:
Maybe deprecate `perm` and `comb`, too, directing folks to `math`? I anticipate that we'll go to 3.8+ this year.

Reply:
How does this look?
 https://github.com/inducer/pytools/compare/36dfffae49cf9dc8eb7446c98f5cd635bfd6f231..501b7a0f5178137e8fad1f4a0fc02f67d416d9ce